### PR TITLE
HTTP/2: Fix some errors reported by h2spec.

### PR DESCRIPTION
**Pull request by buchgr**
Motivation:

h2spec [1] reported 32 issues [2] with Netty's HTTP/2 implementation.

Modifications:

Fixed 9 of those issues. Mostly wrong error codes or added missing validation
code in the frame reader.

Result:

11 fewer errors. h2spec now reports: 70 tests, 48 passed, 1 skipped, 21 failed

[1] https://github.com/summerwind/h2spec
[2] https://github.com/netty/netty/issues/5761

PTL @Scottmitch @nmittler

### CODE REVIEW

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java [hunk 1]
```diff
@@ -599,6 +611,13 @@ private int readPadding(ByteBuf payload) {
         return payload.readUnsignedByte() + 1;
     }
 
+    private void verifyPadding(int padding, int streamId) throws Http2Exception {
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java]: @buchgr remove `streamId` argument ?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java [hunk 2]
```diff
@@ -487,13 +489,13 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
...
-                closeStream(stream, future);
-                break;
+        case HALF_CLOSED_LOCAL:
+        case OPEN:
+            stream.closeLocalSide();
+            break;
+        default:
+            closeStream(stream, future);
+            break;
         }
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: revert formating changes ?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java [hunk 3]
```diff
@@ -604,15 +606,43 @@ protected Http2FrameWriter frameWriter() {
         return encoder().frameWriter();
     }
 
+    /**
+     * Sends a RST_STREAM frame even if we don't know about the stream. This error condition is most likely triggered
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: `{@code RST_STREAM}`

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java [hunk 4]
```diff
@@ -604,15 +606,43 @@ protected Http2FrameWriter frameWriter() {
...
+     * by the first frame of a stream being invalid. That is, there was an error reading the frame, even before we
+     * could create a new stream.
+     */
+    private ChannelFuture resetUnknownStream(final ChannelHandlerContext ctx, int streamId, long errorCode,
+                                             final ChannelPromise promise) {
+        ChannelFuture future = frameWriter().writeRstStream(ctx, streamId, errorCode, promise);
+        if (future.isDone()) {
+            if (!future.isSuccess()) {
+                onConnectionError(ctx, future.cause(), null);
+            }
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: move this code to a new method and reuse in the listener as well

**Comment by buchgr** [comment]
PTAL @normanmaurer

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java [hunk 5]
```diff
@@ -599,6 +611,13 @@ private int readPadding(ByteBuf payload) {
         return payload.readUnsignedByte() + 1;
     }
 
+    private void verifyPadding(int padding) throws Http2Exception {
```
Comment (netkins) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java]: MAJOR Remove this unused method parameter "streamId". [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1172)

Comment (netkins) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java]: MAJOR Remove this unused method parameter "streamId". [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1172)

Comment (netkins) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java]: MAJOR Remove this unused method parameter "streamId". [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1172)

**Comment by nmittler** [comment]
@buchgr LGTM

**Comment by normanmaurer** [comment]
@Scottmitch PTAL as well

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java [hunk 6]
```diff
@@ -566,6 +575,9 @@ private static void readGoAwayFrame(ChannelHandlerContext ctx, ByteBuf payload,
 
     private void readWindowUpdateFrame(ChannelHandlerContext ctx, ByteBuf payload,
             Http2FrameListener listener) throws Http2Exception {
+        if (payloadLength != 4) {
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java]: `Http2CodecUtil.WINDOW_UPDATE_FRAME_LENGTH` exists and uses `INT_FIELD_LENGTH`. Consider using a similar define here or making this check use a common static final variable.

**Comment by Scottmitch** [comment]
1 comment and lgtm

**Comment by buchgr** [comment]
@Scottmitch @normanmaurer @nmittler thank you for the review.

addressed comments and squashed commits.

**Comment by nmittler** [comment]
Still LGTM

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 as 515f8964b42dc34bd42cf0708a11147e2279076f

**Pull request closed without merging**